src/deploy/functions/containerCleaner.ts

```diff
@@ -0,0 +1,160 @@
+    const subdomain = SUBDOMAIN_MAPPING[location] || "us";
+    if (!this.helpers[subdomain]) {
+      this.helpers[subdomain] = new ContainerRegistryHelper(subdomain);
+    }
+    return this.helpers[subdomain];
+  }
+
+  // GCFv1 has the directory structure:
+  // gcf/
+  //  +- <region>/
+  //        +- <uuid>
+  //             +- <hash> (tags: <FuncName>_version-<#>)
+  //             +- cache/ (Only present in first deploy of region)
+  //             |    +- <hash> (tags: latest)
+  //             +- worker/ (Only present in first deploy of region)
+  //                  +- <hash> (tags: latest)
+  //
+  // We'll parallel search for the valid <uuid> and their children
+  // until we find one with the right tag for the function name.
+  // The underlying Helper's caching should make this expensive for
+  // the first function and free for the next functions in the same
+  // region.
+  async cleanupFunction(func: backend.FunctionSpec): Promise<void> {
+    const helper = this.helper(func.region);
+    const uuids = (await helper.ls(`${func.project}/gcf/${func.region}`)).children;
+
+    const uuidTags: Record<string, string[]> = {};
+    const loadUuidTags: Promise<void>[] = [];
+    for (const uuid of uuids) {
+      loadUuidTags.push(
+        (async () => {
+          const path = `${func.project}/gcf/${func.region}/${uuid}`;
+          const tags = (await helper.ls(path)).tags;
+          uuidTags[path] = tags;
+        })()
+      );
+    }
+    await Promise.all(loadUuidTags);
+
+    const extractFunction = /^(.*)_version-\d+$/;
+    const entry = Object.entries(uuidTags).find(([, tags]) => {
+      return tags.find((tag) => {
+        const match = tag.match(extractFunction);
+        return match && match[1] === func.id;
+      });
+    });
+
+    if (!entry) {
+      logger.debug("Could not find image for function", backend.functionName(func));
+      return;
+    }
+    await helper.rm(entry[0]);
+  }
+}
+
+export interface Stat {
+  children: string[];
+  digests: gcr.Digest[];
+  tags: gcr.Tag[];
+}
+
+export class ContainerRegistryHelper {
+  readonly client: gcr.Client;
+  readonly cache: Record<string, Stat> = {};
+
+  constructor(subdomain: string) {
+    this.client = new gcr.Client(subdomain);
+  }
+
+  async ls(path: string): Promise<Stat> {
+    if (!this.cache[path]) {
+      const raw = await this.client.listTags(path);
+      this.cache[path] = {
+        tags: raw.tags,
+        digests: Object.keys(raw.manifest),
+        children: raw.child,
+      };
+    }
+    return this.cache[path];
+  }
+
+  async rm(path: string): Promise<void> {
+    const stat = await this.ls(path);
+    const deleteChildren: Promise<void>[] = [];
+    const recursive = stat.children.map((child) => this.rm(`${path}/${child}`));
+    // Let children ("directories") be cleaned up in parallel while we clean
+    // up the "files" in this location.
+
+    const deleteTags = stat.tags.map((tag) => this.client.deleteTag(path, tag));
```

Comment:
Do we need to delete all tags before starting to delete the images? If not, we could combine these await Promise.all()'s and do them in parallel

Reply:
I'm also a little concerned about how this behaves if just one call fails - for example, if the first Promise.all rejects, this errors and we never try to deleteImages... but we already made our recursive call, so that has started but will get 'cut off' at some arbitrary point.

Reply:
Sounds good to me - this is definitely a tricky one to handle and it seems likely that if one call fails, the rest are high likelihood to as well. I feel better about this strategy tho.